add method for adding standard metrics to existing Resolver

FelipeAdachi opened the pull request:
## Description

It is cumbersome to always have to create a CustomResolver from scratch, even when you might be interested in adding a single metric to the standard resolver. This enables the user to start from an existing resolver and add a Standard Metric according to column name or data type.

## Changes

- Imperative commit or change title (commit ID) <!-- Add commit ID and [GitHub will autolink](https://help.github.com/en/github/writing-on-github/autolinked-references-and-urls). -->
  - Add some bullet points to explain the change
- Next commit or change (commit ID)

## Related

Relates to organization/repo#number

<!-- Reference related commits, issues and pull requests. Type `#` and select from the list. -->

Closes organization/repo#number

<!-- List issues for GitHub to automatically close after merge to default branch. Type `#` and select from the list. See the [GitHub docs on linking PRs to issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) for more. -->

- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

FelipeAdachi committed:
add method for adding standard metrics to existing Resolver

FelipeAdachi committed:
isort

TheMellyBee reviewed (dismissed):
Good idea! Much nicer

jamie256 committed:
Merge branch 'mainline' into dev/felipe/resolver-customization

jamie256 committed:
Merge branch 'mainline' into dev/felipe/resolver-customization

jamie256 committed:
Merge branch 'mainline' into dev/felipe/resolver-customization

FelipeAdachi committed:
add to result instead of replacing it

FelipeAdachi committed:
Merge branch 'dev/felipe/resolver-customization' of https://github.com/whylabs/whylogs into dev/felipe/resolver-customization

jamie256 commented on python/whylogs/core/resolvers.py:
```diff
@@ -11,11 +12,45 @@
+        metric: StandardMetric,
+        column_names: List[String] = [],
+        why_types: List[Union[Fractional, Integral, String]] = [],
+    ):
```
Ok, so this can only add a new `StandardMetric`, that wouldn't already resolve for this type or column name on a resolver, so its rather limited functionally on the `StandardResolver `since you would only be able to add a `StandardMetric`, I guess the main use case here is the `unicode_range `and `condition_count `metrics?

Also its a bit confusing what happens when I add the same `StandardMetric` that would already be resolved, or the same `StandardMetric` type gets added multiple times, perhaps bound to a type and then bound to a column name? It looks like they can both be added but only one will be functional, so maybe it doesn't break things but slows down resolving column schema?

FelipeAdachi replied:
With latest changes, this now expects a SpecResolver, which accepts any kind of Metrics, not just Standard ones.

If the metric to be added is already present in the column's result while resolving the additional specs, it will just skip it.

If both a name and type is passed, the name takes precedence, and will only be added once

jamie256 commented on python/whylogs/core/resolvers.py:
```diff
@@ -11,11 +12,45 @@
+                if column_name == name and additional_metric.metric.name not in result:
+                    additional_result[additional_metric.metric.name] = additional_metric.metric.zero(column_schema.cfg)
+            for target_why_type in additional_metric.why_types:
+                if isinstance(target_why_type, type(why_type)) and additional_metric.metric.name not in result:
```
So this looks like we can only add one ConditionCountMetric? How would I add both a condition count that checks string columns for email pattern, and a condition count that checks the "Fico Score" column is greater than 400? It looks like here the additional_metric.metric.name would just be "condition_count" and limits me to only have one of these?

richard-rogers replied:
A column can only have one `ConditionCountMetric`, but the metric instance can host many conditions (we should probably put a configurable limit on how many ;).

FelipeAdachi replied:
but, right now, there's no easy way to assign different metric instances to different columns, right?

Do we want to change this before merging this PR?

jamie256 commented on python/whylogs/core/resolvers.py:
```diff
@@ -67,6 +110,14 @@ def resolve(self, name: str, why_type: DataType, column_schema: ColumnSchema) ->
+
         return result
 
 
```
Consider adding an EmptyResolver? I do like the HistogramCountingTrackingResolver as a useful base to add onto but it might be better to allow the caller to skip most columns for a very lightweight profile.

richard-rogers replied:
`DeclarativeSchema` is also a handy way to construct a lightweight profile :)

FelipeAdachi replied:
This was added in latest commit!

jamie256 reviewed (commented):
Mostly questions, changes look good.

FelipeAdachi committed:
Merge branch 'mainline' into dev/felipe/resolver-customization

# Conflicts:
#	python/whylogs/core/resolvers.py

FelipeAdachi committed:
change resolver customization to work with SpecResolver instead of AdditionalMetric

changed tests accordingly

FelipeAdachi commented on python/whylogs/core/resolvers.py:
```diff
@@ -11,11 +12,45 @@
+        if not column_names and not why_types:
+            raise ValueError("Either column names or why types must not be empty.")
+        if column_names and why_types:
+            raise ValueError("column_names or why_types should be defined, not both.")
```
yes, now passing both just raises a warning, saying that the name takes precedence.

FelipeAdachi replied:
With latest changes, the validation is made by `DeclarativeResolver` itself. I'm ok with how it works currently. Let me know if we want to change it.

richard-rogers commented on python/whylogs/core/resolvers.py:
```diff
@@ -54,6 +128,13 @@ def resolve(self, name: str, why_type: DataType, column_schema: ColumnSchema) ->
         for m in metrics:
             result[m.name] = m.zero(column_schema.cfg)
+
+        if self.additional_specs:
```
This code seems to be repeated in several places. Consider refactoring into a common implementation.

FelipeAdachi committed:
add EmptyResolver and test on how to use it

richard-rogers commented on python/whylogs/core/resolvers.py:
```diff
@@ -16,11 +16,85 @@
+            Resolver specification that define the metrics to instantiate for matching columns.
+        """
+        if resolver_spec.column_name and resolver_spec.column_type:
+            logger.warning(
```
We should move the `ResolverSpec` validation into `ResolverSpec::__post_init__()`.

FelipeAdachi replied:
Now that I'm using `DeclarativeResolver` to resolve the additional metrics, I removed this validation. I agree that the validation from `DeclarativeResolver` could be moved to `ResolverSpec::__post_init__()`.

I'm ok with doing this in a follow-up, though. lmk if that's ok

richard-rogers replied:
It will be a slightly breaking change, so it would be good to fix it before `DeclarativeSchema` gets released Tuesday :(

FelipeAdachi committed:
- Using DeclarativeResolver to resolve additional specs
- Make resolve_additional_specs an abstract method in Resolver and moving implementation to each subclass
- improving docstrings
- refactoring to avoid code duplication in resolve_additional_specs

FelipeAdachi commented on python/whylogs/core/resolvers.py:
```diff
@@ -16,11 +16,85 @@
+        for resolver_spec in self.additional_specs:
+            col_name, col_type = resolver_spec.column_name, resolver_spec.column_type
+
+            if (col_name and col_name == name) or (col_name is None and isinstance(why_type, col_type)):  # type: ignore
```
Absolutely!

> Perhaps refactor to share code with DeclarativeSchema?

With latest changes, I'm now using DeclarativeResolver to resolve the additional metrics.

> Also, maybe this should be an abstract method in the Resolver base class?

I changed `resolve_additional_specs` to be abstract in Resolver Base Class, and called `super().resolve_additional_specs` in each subclass. Is that what you meant?

But wouldn't this be a breaking change, since that requires us to change every custom resolver created on the examples/tests to add that method?

FelipeAdachi committed:
flake

FelipeAdachi committed:
implement add_resolver_spec in abstract class and call super in each subclass

richard-rogers commented on python/tests/core/test_resolver.py:
```diff
@@ -0,0 +1,159 @@
+    )
+
+    resolver.add_resolver_spec(legs_not_4_spec)
+    resolver.add_resolver_spec(weights_not_4_2_spec)
```
Maybe change `add_resolver_spec` to `add_resolver` or `add_metric`? Maybe pass in `Union[ResolverSpec, List[ResolverSpec]]` ?

FelipeAdachi commented:
Closing this as #1016 is an updated version of this PR
